keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -304,31 +312,43 @@ class ResourceActuator(
       else -> SystemException(this)
     } as SpinnakerException
 
-  private suspend fun <T : Any> ResourceHandler<*, T>.resolve(resource: Resource<ResourceSpec>): Pair<T, T?> =
-    supervisorScope {
-      val desired = async {
-        try {
-          desired(resource)
-        } catch (e: ResourceCurrentlyUnresolvable) {
-          throw e
-        } catch (e: Throwable) {
-          throw CannotResolveDesiredState(resource.id, e)
+  private suspend fun <T : Any> ResourceHandler<*, T>.resolve(resource: Resource<ResourceSpec>): Pair<T, T?> {
+    val isKotlin = this.javaClass.isKotlinClass
+    return supervisorScope {
+      val desired = if (isKotlin) {
```

Comment:
This now checks if the plugin is a Kotlin class before using `async` to call `desired`, and falls back to calling `desiredAsync` (Java-friendly version) below if it's not. Same thing for `current` immediately below.

Reply:
While this code works, I think the semantics would more clearly match the code if the `ResourceHandler` interface instead specified an enum that you could switch on to choose `desired` versus `desiredAsync`, something like this:

```kotlin
//
// ResourceHandler.kt
//
enum class Implementation {
 COROUTINE, ASYNC
}

interface ResourceHandler<S : ResourceSpec, R : Any> : SpinnakerExtensionPoint {
 val implementation : Implementation
}

//
// ResourceActuator.kt
//
private suspend fun <T : Any> ResourceHandler<*, T>.resolve(resource: Resource<ResourceSpec>) =
 supervisorScope {
  when(implementation) {
   COROUTINE -> { ... }
   ASYNC -> {...}
 }
}
```

Reply:
Oooh, that's good! I'll do that! Thanks for the suggestion.

Reply:
Yeah, I agree that we should not have two separate checks. Ultimately we have to just trust that the plugin implementer is doing the right thing. I suggested the enum approach because I thought it would make the intent a little clearer in the ResourceActuator code, but it's really just an aesthetic judgement.